src/core/dbus-unit.c

```diff
@@ -722,6 +723,66 @@ int bus_unit_method_clean(sd_bus_message *message, void *userdata, sd_bus_error
         return sd_bus_reply_method_return(message, NULL);
 }
 
+int bus_unit_method_freeze(sd_bus_message *message, void *userdata, sd_bus_error *error) {
+        Unit *u = userdata;
+        int r;
+
+        assert(message);
+        assert(u);
+
+        if (u->pending_freezer_message) {
+                /* Operation that was requested previously is still in progress */
+                unit_add_to_dbus_queue(u);
+                return sd_bus_error_setf(error, SD_BUS_ERROR_FAILED, "Previously requested freezer operation for unit '%s' is still in progress.", u->id);
+        }
+
+        r = mac_selinux_unit_access_check(u, message, "stop", error);
+        if (r < 0)
+                return r;
+
+        r = sd_bus_message_new_method_return(message, &u->pending_freezer_message);
+        if (r < 0)
+                return r;
+
+        r = unit_freeze(u);
+        if (r == -EOPNOTSUPP)
+                return sd_bus_error_setf(error, SD_BUS_ERROR_NOT_SUPPORTED, "Unit '%s' does not supporting freezing.", u->id);
```

Comment:
This is misleading -- this implies that this unit will never support freezing, but it's not necessarily the case. It could also be that the cgroup is not set up properly yet (from unit_cgroup_freezer_action), that we're not on unified hierarchy (so it's not unit-specific), etc...

Either that or consider EOPNOTSUPP vs. EBUSY (I'll leave you to think about semantics here, but maybe it at least allows to distinguish the specificity of the issue)

Reply:
Well, the message says that the unit doesn't support freezing at the time of the request. I don't see there any future implication. It is always true for every unit on cgroup v1. With v2 we support freezing only when unit type has freezing implemented. Also units cgroup must be already in place at the time we get Freeze/Thaw request since the `u->cgroup_path` is established before we even fork-off the service binary (https://github.com/systemd/systemd/blob/master/src/core/service.c#L1458)

However, I should add the case for EBUSY. I feel like the error handling for freezing should be overall similar to what we do in `bus_unit_method_clean()`.

Reply:
Reference to `service.c` is not entirely relevant since the freezing is supported also by slices and scopes. But I think that you can't have a unit that is active (we support freezing only active units) and cgroup that is not realized.